src/Querys/BatchQueryResult.php

```diff
@@ -0,0 +1,305 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Db\Querys;
+
+use Yiisoft\Db\Data\DataReader;
+use Yiisoft\Db\Drivers\Connection;
+use Yiisoft\Db\Exceptions\Exception;
+
+/**
+ * BatchQueryResult represents a batch query from which you can retrieve data in batches.
+ *
+ * You usually do not instantiate BatchQueryResult directly. Instead, you obtain it by
+ * calling {@see Query::batch()} or {@see Query::each()}. Because BatchQueryResult implements the {@see \Iterator}
+ * interface, you can iterate it to obtain a batch of data in each iteration. For example,
+ *
+ * ```php
+ * $query = (new Query)->from('user');
+ * foreach ($query->batch() as $i => $users) {
+ *     // $users represents the rows in the $i-th batch
+ * }
+ * foreach ($query->each() as $user) {
+ * }
+ * ```
+ */
+class BatchQueryResult implements \Iterator
+{
+    /**
+     * @var Connection|null the DB connection to be used when performing batch query.
+     * If null, the "db" application component will be used.
+     */
+    private ?Connection $db;
```

Comment:
Default value is not defined

Reply:
I think that it and `$query` should be required in constructor

Reply:
For now I have them with setters, I already defined the default values with null, I think that in this case having the cleanest constructors, it will make the classes work better with the dependencies, anyway once this PR has been implemented, these improvements can be implemented.